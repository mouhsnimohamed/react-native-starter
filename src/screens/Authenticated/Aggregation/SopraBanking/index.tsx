import React from 'react';
import { WebView } from 'react-native-webview';
import { StackNavigationProp } from '@react-navigation/stack';
import { RootStackParamList } from 'navigations/RootNavigation';
import { Root, Spinner } from 'components';
import styled from 'styled-components/native';
import i18n from 'translations';
import SopraBankingResponse from '../components/SopraBankingResponse';
import { useTypedSelector } from 'redux/root.reducers';
import { aggregationSelector } from 'modules/aggregation/aggregation.selectors';
import { WebViewProgressEvent } from 'react-native-webview/lib/WebViewTypes';
import getParams from 'helpers/extractUrlParams';
import { useDispatch } from 'react-redux';
import { confirmConsentAction } from 'modules/aggregation/aggregation.actions';

interface SopraBankingProps {
  navigation: StackNavigationProp<RootStackParamList, 'SopraBanking'>;
}
const SopraBanking: React.FC<SopraBankingProps> = ({ navigation }) => {
  const { sopraBankingURI } = useTypedSelector(aggregationSelector);

  const dispatch = useDispatch();
  const RESPONSE_URL = 'https://appmobile-rec.bmce-euroservices.com';

  const changedUrlListener = ({ nativeEvent }: WebViewProgressEvent) => {
    const { url } = nativeEvent;
    if (url.startsWith(RESPONSE_URL)) {
      const params = getParams(url);
      if (params.credentialsId && params.state) {
        dispatch(confirmConsentAction(params));
        navigation.navigate('SopraBankingResponseScreen', {
          variant: 'success',
        });
      } else {
        navigation.navigate('SopraBankingResponseScreen', {
          variant: 'error',
        });
      }
    }
  };

  return (
    <Root>
      <WebViewHeight>
        <WebView
          source={{
            uri: `${sopraBankingURI}`,
          }}
          automaticallyAdjustContentInsets={false}
          javaScriptEnabledAndroid={true}
          domStorageEnabled={true}
          javaScriptCanOpenWindowsAutomatically={true}
          originWhitelist={['*']}
          onLoadProgress={changedUrlListener}
          startInLoadingState={true}
          renderLoading={() => <Spinner isLoading={true} />}
          renderError={() => (
            <ErrotContainer>
              <SopraBankingResponse
                variant="error"
                title={i18n.t('aggregations.webview.errorOnLoad')}
                CTATitle={i18n.t('aggregations.webview.retry')}
                onPress={() => navigation.goBack()}
              />
            </ErrotContainer>
          )}
        />
      </WebViewHeight>
    </Root>
  );
};

export default SopraBanking;

const WebViewHeight = styled.View`
  flex: 1;
`;

const ErrotContainer = styled.View`
  bottom: 0;
  flex: 1;
  justify-content: center;
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
`;
