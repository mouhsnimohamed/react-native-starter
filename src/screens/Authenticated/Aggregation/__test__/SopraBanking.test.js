import SopraBanking from '../SopraBanking';
import React from 'react';
import { render, cleanup } from 'utils/test/render.utils';
import {
  AGGREGATION_REDUCER,
  aggregationReducer,
  initState,
} from 'modules/aggregation/aggregation.reducer';

afterEach(cleanup);

describe('<SopraBanking />', () => {
  it('render SopraBanking WebView correctly', () => {
    const rendered = render(<SopraBanking />, {
      reducers: {
        [AGGREGATION_REDUCER]: () =>
          aggregationReducer(
            {
              ...initState,
              sopraBankingURI: 'https://recettebes.bankofafrica.ma',
            },
            {},
          ),
      },
    }).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});
