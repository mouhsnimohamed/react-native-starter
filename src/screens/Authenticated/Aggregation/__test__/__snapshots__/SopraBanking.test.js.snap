// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SopraBanking /> render SopraBanking WebView correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          undefined,
        ]
      }
    >
      <RNCWebView
        automaticallyAdjustContentInsets={false}
        cacheEnabled={true}
        domStorageEnabled={true}
        injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
        injectedJavaScriptForMainFrameOnly={true}
        javaScriptCanOpenWindowsAutomatically={true}
        javaScriptEnabled={true}
        javaScriptEnabledAndroid={true}
        messagingEnabled={false}
        onContentProcessDidTerminate={[Function]}
        onHttpError={[Function]}
        onLoadProgress={[Function]}
        onLoadingError={[Function]}
        onLoadingFinish={[Function]}
        onLoadingProgress={[Function]}
        onLoadingStart={[Function]}
        onMessage={[Function]}
        onShouldStartLoadWithRequest={[Function]}
        source={
          Object {
            "uri": "https://recettebes.bankofafrica.ma",
          }
        }
        startInLoadingState={true}
        style={
          Array [
            Object {
              "flex": 1,
              "overflow": "hidden",
            },
            Object {
              "backgroundColor": "#ffffff",
            },
            undefined,
          ]
        }
        useSharedProcessPool={true}
      />
      <Modal
        animationType="none"
        hardwareAccelerated={false}
        transparent={true}
        visible={true}
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "rgba(255,255,255,0.3)",
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "justifyContent": "center",
              },
            ]
          }
        >
          <ActivityIndicator
            animating={true}
            color="#003D7C"
            hidesWhenStopped={true}
            size="large"
          />
        </View>
      </Modal>
    </View>
  </View>
</View>
`;
